Improve Quantity Formatting Documentation (#7045)
iTwin/itwinjs-core: core/quantity/README.md

# @itwin/core-quantity

The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js reference documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).

> Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/quantity/index.md) explaining quantity formatting and its basic concepts.

Reviewer: Do you mean https://www.itwinjs.org/learning/frontend/quantityformatting ?

> Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/quantity/index.md) explaining quantity formatting and its basic concepts.

Reviewer: ```suggestion
Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/quantity) explaining quantity formatting and its basic concepts.
```

# @itwin/core-quantity

- [@itwin/core-quantity](#itwincore-quantity)
  - [Terms and Concepts](#terms-and-concepts)
    - [Common Terms](#common-terms)
    - [Concepts](#concepts)
      - [Units Provider](#units-provider)
      - [Unit Conversion](#unit-conversion)
    - [Examples of Usage](#examples-of-usage)
      - [Numeric Format](#numeric-format)
      - [Composite Format](#composite-format)
      - [Parsing Values](#parsing-values)
      - [Mathematical Operation Parsing](#mathematical-operation-parsing)
  - [Contributing](#contributing)
- [Licensing](#licensing)

The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing.

## Terms and Concepts

### Common Terms

> - [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.

Author: Instead of using $quantity here and everywhere else in this file, should I link straight to the line of code of the relevant interfaces and methods? Assuming this readme isn't part of the docs site

Reviewer 2: I'm not familiar with using variables in markdown like this. Where do we define $quantity? The github preview for the file turns this into dead links.
The idea of putting line of code links here does not appeal to me. I've never seen that being done in abstract code documentation. People will read this .md to understand concepts. Also code links could easily go out of sync.

Author: As an example, see this [README.md](https://github.com/iTwin/itwinjs-core/blob/master/docs/learning/frontend/QuantityFormatting.md?plain=1), and this [link](https://www.itwinjs.org/learning/frontend/quantityformatting/) shows how it is displayed in the docs site. So, these $<XXX_packagename> links really only make sense in the context of our itwinjs.org docs site. So I think I will remove the $quantity and everywhere else... I agree with the code links, I won't introduce them. Instead, I'll put a link at the top of the opening description, linking to the docs site's API reference for core-quantity

Author: okay, added

# @itwin/core-quantity

- [@itwin/core-quantity](#itwincore-quantity)
  - [Terms and Concepts](#terms-and-concepts)
    - [Common Terms](#common-terms)
    - [Concepts](#concepts)
      - [Units Provider](#units-provider)
      - [Unit Conversion](#unit-conversion)
    - [Examples of Usage](#examples-of-usage)
      - [Numeric Format](#numeric-format)
      - [Composite Format](#composite-format)
      - [Parsing Values](#parsing-values)
      - [Mathematical Operation Parsing](#mathematical-operation-parsing)
  - [Contributing](#contributing)
- [Licensing](#licensing)

> The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js references documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).

Reviewer: ```suggestion
The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js reference documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).
```

Author: addressed

> Also check the [iTwin.js learnings documentation](https://www.itwinjs.org/learning/frontend/quantityformatting/#quantity-package) explaining quantity formatting and it's basic concepts.

Reviewer: ```suggestion
Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/frontend/quantityformatting/#quantity-package) explaining quantity formatting and its basic concepts.
```

Author: addressed

## Terms and Concepts

### Common Terms

> - `Unit`/`UnitProps` - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.

Author: addressed

> - `UnitsProvider` - An interface that locates the UnitProps for a unit given name or label. This interface also provides methods for `UnitConversion` to allow converting from one unit to another.

Reviewer: ```suggestion
- `UnitsProvider` - An interface that locates the `UnitProps` for a unit given name or label. This interface also provides methods for `UnitConversion` to allow converting from one unit to another.
```

Author: addressed

- Unit Family/`Phenomenon` - A physical quantity that can be measured (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel.
- Format/FormatProp - The display format for the quantity value. For example, an angle may be persisted in radians but formatted and shown to user in degrees.
  - CompositeValue - An addition to the format specification that allows the explicit specification of a unit label, it also allows the persisted value to be displayed as up to 4 sub-units. Typical multi-unit composites are used to display `feet'-inches"` and `degree°minutes'seconds"`.

> - `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the UnitsProvider during the formatting process.

Reviewer: ```suggestion
- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the `UnitsProvider` during the formatting process.
```

Author: addressed

> - `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the UnitsProvider and also done to allow a user to enter `43in` even when in "metric" unit system and have the string properly converted to meters.

Reviewer: ```suggestion
- `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the `UnitsProvider` and to allow users to input quantities in different unit systems than specified. For instance, if a metric unit system is set, a user could enter `43in` and have the result properly converted to meters. 
```

Author: addressed

> - `Formatter` - A class that holds methods to format a quantity value into a text string. Given a FormatterSpec object - containing one or many Unit definitions each with their own Unit Conversion info and a Format passed in - and a single `magnitude` number, the Formatter can convert that number into a text string, adhering to the properties passed to the formatTraits of a Format.

Reviewer: ```suggestion
- `Formatter` - A class that holds methods to format a quantity value into a text string. Given a `FormatterSpec` object—which includes one or more unit definitions, each with their own conversion information and a specified format—and a single magnitude number, the `Formatter` can convert this number into a text string, adhering to the properties specified in the `formatTraits`.
```

Author: addressed

> - `Parser` - A class that holds methods to parse a text string into a single number. Given a ParserSpec object containing a Format's Units and their Unit Conversions, as well as an input string, the Parser can either return an object `QuantityParseResult` that contains the magnitude of type `number`, or an object `ParseQuantityError`.

Reviewer: ```suggestion
- `Parser` - A class that holds methods to parse a text string into a single number. Given a `ParserSpec` object containing a Format's Units and their Unit Conversions, as well as an input string, the Parser can either return an object `QuantityParseResult` that contains the magnitude of type `number`, or an object `ParseQuantityError`.
```

Author: addressed

### Concepts

#### Units Provider

> To appropriately parse and output formatted values, A units provider is used to define all available units and provides conversion factors between units. There are a couple implementations of the UnitsProvider across iTwin.js:

Reviewer: ```suggestion
To appropriately parse and output formatted values, a units provider is used to define all available units and provides conversion factors between units. There are several implementations of the UnitsProvider across iTwin.js:
```

Author: addressed

The `BasicUnitsProvider` holds many common units and their conversions between each other.

> The `SchemaUnitProvider` is another example, used to load unit definitions of schemas from an iModel. This holds more extensive units through the Units schema, while also allowing users to define their own units.

Reviewer: ```suggestion
The `SchemaUnitProvider` is used to load unit definitions of schemas from an iModel. This holds more extensive units through the Units schema, while also allowing users to define their own units.
```

Author: addressed

The `AlternateUnitLabelsProvider` interface allows users to specify a set of alternate labels which may be encountered during parsing of strings. By default only the input unit label and the labels of other units in the same Unit Family/Phenomenon, as well as the label of units in a Composite format are used.

#### Unit Conversion

> Unit conversion is performed through `UnitConversionSpec`. These objects are generated by a UnitsProvider, with the implementation determined by each specific provider. During initialization, a ParserSpec/FormatterSpec can ask for UnitConversionSpec objects provided via the UnitsProvider. During parsing and formatting, the specification will retrieve the UnitConversionSpec between the source and destination units to apply the unit conversion.

Reviewer: ```suggestion
Unit conversion is performed through a `UnitConversionSpec`. These objects are generated by a `UnitsProvider`, with the implementation determined by each specific provider. During initialization, a `ParserSpec` or `FormatterSpec` can ask for `UnitConversionSpec` objects provided via the `UnitsProvider`. During parsing and formatting, the specification will retrieve the `UnitConversionSpec` between the source and destination units to apply the unit conversion.
```

> <!-- #### How a value is formatted

Reviewer: Stray comments? TODOs?

Author: Yes, stray comments - I've removed them

    // result in formattedValue of 3'-3 3/8"
```

</details>

#### Parsing Values

<details>
  <summary>Example Code:</summary>

```ts
  // define output unit and also used to determine the unit family used during parsing
  const outUnit = await unitsProvider.findUnitByName("Units.M");

  const formatData = {
    composite: {
      includeZero: true,
      spacer: "-",
      units: [{ label: "'", name: "Units.FT" }, { label: "\"", name: "Units.IN" }],
    },
    formatTraits: ["keepSingleZero", "showUnitLabel"],
    precision: 8,
    type: "Fractional",
    uomSeparator: "",
  };

  // generate a Format from FormatProps used to determine possible labels
  const format = new Format("test");
  await format.fromJSON(unitsProvider, formatData);

  const inString = "2FT 6IN";

  // create the parserSpec spec which will hold all unit conversions from possible units to the output unit
  const parserSpec = await ParserSpec.create(format, unitsProvider, outUnit, unitsProvider);
  const parseResult = parserSpec.parseToQuantityValue(inString);
  //  parseResult.value 0.762  (value in meters)
```

</details>

#### Mathematical Operation Parsing

> The quantity formatter supports parsing mathematical operations. The operation is solved, formatting every values present, according to the specified format. This makes it possible to process several different units at once.

Reviewer: ```suggestion
The quantity formatter supports parsing mathematical operations. The operation is solved, formatting each value present, according to the specified format. This makes it possible to process several different units at once.
```

Author: addressed

Copyright © Bentley Systems, Incorporated. All rights reserved. See LICENSE.md for license terms and full copyright notice.

The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing.

## Terms and Concepts

### Common Terms

- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.

> - [UnitsProvider]($quantity) - A class that will also locate the UnitProps for a unit given name or label. This class will also provide a [UnitConversion]($quantity) to convert from one unit to another.

Reviewer 3: you say a UnitsProvider 'will also locate' and 'will also provide' ... seems wrong for the first sentence to say 'will also' because nothing comes before it.

Author: Fixed here and in the other QuantityFormatting.md

> - Unit Family/[Phenomenon]($ecschema-metadata) - The physical quantity that this unit measures (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.

Reviewer 3: what is 'this unit'?  Should this be written as 'A physical quantity that can be measured'

Author: that actually was copy pasted from the other existing quantity formatting doc, I'll fix this here and in the other spot

Author: Okay, fixed

> - Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel. While there is **NO** explicit flag or property to denote a persistence unit, those units are typically a part of the `Units.SI` [UnitSystem]($ecschema-metadata). The unit conversion of any persistence unit should both have its numerator and denominator set to 1.

Reviewer 3: This could be confusing: 'any persistence unit should both have it's numerator and denominator set to 1', I'm pretty sure that's true but so many other units have this true that I'm not sure it matters.  

Also this is mixing levels.  Persistence Unit is the unit used to persist the value, what we choose as the persistence unit is based on our schema standards.  There is nothing in core quantity that mandates that persistence unit be SI.

Author: okay, so maybe I'll put the blurb about persistence unit in the learning doc in core-frontend?

Author: QuantityFormatting.md already had a sentence on persistence unit from Bill, so I switched to using that
